From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nima Vasseghi <8564939+nima2007@users.noreply.github.com>
Date: Mon, 20 Jul 2026 16:16:18 -0700
Subject: [PATCH] mmap hack: forward-port for macOS 27 support

Forward-port the mmap changes to current idevicerestore, including modern IMG4 component tag handling.

Based-on-patch-by: Cosmo Chou <cosmo.chou@quantatw.com>
---
 src/common.c         | 14 ++-----
 src/common.h         | 13 +++++++
 src/dfu.c            | 14 ++++++-
 src/idevicerestore.c | 13 ++++---
 src/img4.c           | 84 ++++++++++++++++++++++------------------
 src/ipsw.c           | 73 +++++++++++++++++++----------------
 src/ipsw.h           |  2 +-
 src/recovery.c       |  2 +-
 src/restore.c        | 92 ++++++++++++++++++++++++++------------------
 9 files changed, 178 insertions(+), 129 deletions(-)

diff --git a/src/common.c b/src/common.c
index 70f0742..a1dde2c 100644
--- a/src/common.c
+++ b/src/common.c
@@ -126,18 +126,10 @@ int read_file(const char* filename, void** data, size_t* size) {
 	}
 	length = fst.st_size;
 
-	buffer = (char*) malloc(length);
-	if (buffer == NULL) {
-		logger(LL_ERROR, "Out of memory\n");
-		fclose(file);
-		return -1;
-	}
-	bytes = fread(buffer, 1, length, file);
+	buffer = (char *)mmap(0, length, PROT_READ, MAP_PRIVATE, fileno(file), 0);
 	fclose(file);
-
-	if (bytes != length) {
-		logger(LL_ERROR, "Unable to read entire file\n");
-		free(buffer);
+	if (buffer == MAP_FAILED) {
+		logger(LL_ERROR, "%s: mmap failed for %s: %s\n", __func__, filename, strerror(errno));
 		return -1;
 	}
 
diff --git a/src/common.h b/src/common.h
index 3bb648a..8757a9f 100644
--- a/src/common.h
+++ b/src/common.h
@@ -39,6 +39,9 @@ extern "C" {
 #include <libirecovery.h>
 #include <libimobiledevice-glue/thread.h>
 
+#include <errno.h>
+#include <sys/mman.h>
+
 #include "idevicerestore.h"
 #include "log.h"
 
@@ -69,6 +72,16 @@ extern "C" {
 
 #define USER_AGENT_STRING "InetURL/1.0"
 
+#define free_mmap(data, size, fail_return) \
+if (data) { \
+	if (!size) { \
+		free(data); \
+	} else if (munmap(data, size)) { \
+		logger(LL_ERROR, "%s: munmap failed: %s (%d)\n", __func__, strerror(errno), errno); \
+		return fail_return; \
+	} \
+}
+
 struct dfu_client_t;
 struct normal_client_t;
 struct restore_client_t;
diff --git a/src/dfu.c b/src/dfu.c
index 6ba23e7..e069dae 100644
--- a/src/dfu.c
+++ b/src/dfu.c
@@ -127,6 +127,7 @@ int dfu_send_buffer(struct idevicerestore_client_t* client, const void* buffer,
 int dfu_send_component(struct idevicerestore_client_t* client, plist_t build_identity, const char* component)
 {
 	char* path = NULL;
+	bool is_mmap = true;
 
 	// Use a specific TSS ticket for the Ap,LocalPolicy component
 	plist_t tss = client->tss;
@@ -142,6 +143,7 @@ int dfu_send_component(struct idevicerestore_client_t* client, plist_t build_ide
 		component_data = malloc(sizeof(lpol_file));
 		component_size = sizeof(lpol_file);
 		memcpy(component_data, lpol_file, component_size);
+		is_mmap = false;
 	} else {
 		if (tss) {
 			if (tss_response_get_path_by_entry(tss, component, &path) < 0) {
@@ -170,10 +172,18 @@ int dfu_send_component(struct idevicerestore_client_t* client, plist_t build_ide
 
 	if (personalize_component(client, component, component_data, component_size, tss, &data, &size) < 0) {
 		logger(LL_ERROR, "Unable to get personalized component: %s\n", component);
-		free(component_data);
+		if (is_mmap) {
+			free_mmap(component_data, component_size, -1);
+		} else {
+			free(component_data);
+		}
 		return -1;
 	}
-	free(component_data);
+	if (is_mmap) {
+		free_mmap(component_data, component_size, -1);
+	} else {
+		free(component_data);
+	}
 	component_data = NULL;
 
 	if (!client->image4supported && client->build_major > 8 && !(client->flags & FLAG_CUSTOM) && !strcmp(component, "iBEC")) {
diff --git a/src/idevicerestore.c b/src/idevicerestore.c
index 11d7ec4..b0b16a2 100644
--- a/src/idevicerestore.c
+++ b/src/idevicerestore.c
@@ -237,7 +237,7 @@ static int load_version_data(struct idevicerestore_client_t* client)
 
 	client->version_data = NULL;
 	plist_from_xml(verbuf, verlen, &client->version_data);
-	free(verbuf);
+	free_mmap(verbuf, verlen, -1);
 
 	if (!client->version_data) {
 		remove(version_xml);
@@ -492,7 +492,7 @@ int idevicerestore_start(struct idevicerestore_client_t* client)
 		}
 		dfu_client_free(client);
 
-		free(wtftmp);
+		free_mmap(wtftmp, wtfsize, -1);
 
 		cond_wait_timeout(&client->device_event_cond, &client->device_event_mutex, 10000);
 		if (client->mode != MODE_DFU || (client->flags & FLAG_QUIT)) {
@@ -804,6 +804,7 @@ int idevicerestore_start(struct idevicerestore_client_t* client)
 		if (request == NULL) {
 			plist_free(parameters);
 			logger(LL_ERROR, "Unable to create TSS request\n");
+			free_mmap(uarp_buf, uarp_size, -1);
 			return -1;
 		}
 		plist_dict_merge(&request, parameters);
@@ -814,6 +815,7 @@ int idevicerestore_start(struct idevicerestore_client_t* client)
 		plist_free(request);
 		if (response == NULL) {
 			logger(LL_ERROR, "Unable to send TSS request\n");
+			free_mmap(uarp_buf, uarp_size, -1);
 			return -1;
 		}
 		logger(LL_INFO, "Received USBPortController1,Ticket\n");
@@ -823,10 +825,11 @@ int idevicerestore_start(struct idevicerestore_client_t* client)
 		size_t ace3bin_size = 0;
 		if (ace3_create_binary(uarp_buf, uarp_size, pdfu_bdid, prev, response, &ace3bin, &ace3bin_size) < 0) {
 			logger(LL_ERROR, "Could not create Ace3Binary\n");
+			free_mmap(uarp_buf, uarp_size, -1);
 			return -1;
 		}
 		plist_free(response);
-		free(uarp_buf);
+		free_mmap(uarp_buf, uarp_size, -1);
 
 		if (client->flags & FLAG_KEEP_PERS) {
 			write_file("Ace3Binary", ace3bin, ace3bin_size);
@@ -933,7 +936,7 @@ int idevicerestore_start(struct idevicerestore_client_t* client)
 				}
 				tok = strtok(NULL, "\r\n");
 			}
-			free(fmanifest);
+			free_mmap(fmanifest, msize, -1);
 
 			manifest = plist_new_dict();
 
@@ -1660,7 +1663,7 @@ void idevicerestore_client_free(struct idevicerestore_client_t* client)
 		free(client->cache_dir);
 	}
 	if (client->root_ticket) {
-		free(client->root_ticket);
+		munmap(client->root_ticket, client->root_ticket_len);
 	}
 	if (client->build_manifest) {
 		plist_free(client->build_manifest);
diff --git a/src/img4.c b/src/img4.c
index 732ef69..5b97ac2 100644
--- a/src/img4.c
+++ b/src/img4.c
@@ -415,6 +415,7 @@ int img4_stitch_component(const char* component_name, const void* component_data
 	unsigned char* p;
 	unsigned char* blob = NULL;
 	unsigned int blob_size = 0;
+	unsigned char* p_data;
 
 	if (!component_name || !component_data || component_size == 0 || !tss_response || !img4_data || !img4_size) {
 		return -1;
@@ -426,44 +427,6 @@ int img4_stitch_component(const char* component_name, const void* component_data
 	}
 
 	logger(LL_INFO, "Personalizing IMG4 component %s...\n", component_name);
-	/* first we need check if we have to change the tag for the given component */
-	const void *tag = asn1_find_element(1, ASN1_IA5_STRING, component_data);
-	if (tag) {
-		logger(LL_DEBUG, "Tag found\n");
-		const char* matches[] = {
-			"RestoreKernelCache",
-			"RestoreDeviceTree",
-			"RestoreSEP",
-			"RestoreLogo",
-			"RestoreTrustCache",
-			"RestoreDCP",
-			"Ap,RestoreDCP2",
-			"Ap,RestoreTMU",
-			"Ap,RestoreCIO",
-			"Ap,DCP2",
-			"Ap,RestoreSecureM3Firmware",
-			"Ap,RestoreSecurePageTableMonitor",
-			"Ap,RestoreTrustedExecutionMonitor",
-			"Ap,RestorecL4",
-			NULL
-		};
-		int i = 0;
-		while (matches[i]) {
-			if (!strcmp(matches[i], component_name)) {
-				const char* comptag = _img4_get_component_tag(component_name);
-				if (comptag) {
-					memcpy((void*)tag, comptag, 4);
-				} else {
-					logger(LL_WARNING, "Cannot find tag for component '%s'\n", component_name);
-				}
-				break;
-			}
-			i++;
-		}
-	} else {
-		logger(LL_ERROR, "Personalization failed for component '%s': Tag not found\n", component_name);
-		return -1;
-	}
 
 	// check if we have a *-TBM entry for the given component
 	unsigned char *additional_data = NULL;
@@ -707,6 +670,7 @@ int img4_stitch_component(const char* component_name, const void* component_data
 	p += magic_header_size;
 	memcpy(p, IMG4_MAGIC, IMG4_MAGIC_SIZE);
 	p += IMG4_MAGIC_SIZE;
+	p_data = p;
 	memcpy(p, component_data, component_size);
 	p += component_size;
 	memcpy(p, blob_header, blob_header_size);
@@ -718,6 +682,50 @@ int img4_stitch_component(const char* component_name, const void* component_data
 		p += additional_size;
 	}
 
+	/* first we need check if we have to change the tag for the given component */
+	const void *tag = asn1_find_element(1, ASN1_IA5_STRING, p_data);
+	if (tag) {
+		logger(LL_DEBUG, "Tag found\n");
+		const char* matches[] = {
+			"RestoreKernelCache",
+			"RestoreDeviceTree",
+			"RestoreSEP",
+			"RestoreLogo",
+			"RestoreTrustCache",
+			"RestoreDCP",
+			"Ap,RestoreDCP2",
+			"Ap,RestoreTMU",
+			"Ap,RestoreCIO",
+			"Ap,DCP2",
+			"Ap,RestoreSecureM3Firmware",
+			"Ap,RestoreSecurePageTableMonitor",
+			"Ap,RestoreTrustedExecutionMonitor",
+			"Ap,RestorecL4",
+			NULL
+		};
+		int i = 0;
+		while (matches[i]) {
+			if (!strcmp(matches[i], component_name)) {
+				const char* comptag = _img4_get_component_tag(component_name);
+				if (comptag) {
+					memcpy((void*)tag, comptag, 4);
+				} else {
+					logger(LL_WARNING, "Cannot find tag for component '%s'\n", component_name);
+				}
+				break;
+			}
+			i++;
+		}
+	} else {
+		logger(LL_ERROR, "Personalization failed for component '%s': Tag not found\n", component_name);
+		free(outbuf);
+		free(magic_header);
+		free(blob_header);
+		free(img4header);
+		free(additional_data);
+		return -1;
+	}
+
 	*img4_data = outbuf;
 	*img4_size = (p - outbuf);
 
diff --git a/src/ipsw.c b/src/ipsw.c
index cebfa49..3e37b51 100644
--- a/src/ipsw.c
+++ b/src/ipsw.c
@@ -122,7 +122,7 @@ int ipsw_print_info(const char* path)
 
 	plist_t manifest = NULL;
 	plist_from_memory(plist_buf, plist_len, &manifest, NULL);
-	free(plist_buf);
+	free_mmap(plist_buf, plist_len, -1);
 
 	plist_t val;
 
@@ -633,6 +633,9 @@ int ipsw_extract_to_memory(ipsw_archive_t ipsw, const char* infile, void** pbuff
 	}
 
 	if (ipsw->zip) {
+		logger(LL_ERROR, "zip mode unsupported with mmap\n");
+		return -1;
+
 		int err = 0;
 		struct zip *zip = zip_open(ipsw->path, 0, &err);
 		if (zip == NULL) {
@@ -718,43 +721,41 @@ int ipsw_extract_to_memory(ipsw_archive_t ipsw, const char* infile, void** pbuff
 			free(filepath);
 			return -1;
 		}
-		buffer = (unsigned char*)malloc(size+1);
-		if (buffer == NULL) {
-			logger(LL_ERROR, "Out of memory\n");
-			free(filepath);
-			return -1;
-		}
 
 #ifndef WIN32
 		if (S_ISLNK(fst.st_mode)) {
+			buffer = (unsigned char*)malloc(size+1);
+			if (buffer == NULL) {
+				logger(LL_ERROR, "Out of memory\n");
+				free(filepath);
+				return -1;
+			}
 			if (readlink(filepath, (char*)buffer, size) < 0) {
 				logger(LL_ERROR, "%s: readlink failed for %s: %s\n", __func__, filepath, strerror(errno));
 				free(filepath);
 				free(buffer);
 				return -1;
 			}
+			buffer[size] = '\0';
+			size = 0;
 		} else {
 #endif
 			FILE *f = fopen(filepath, "rb");
 			if (!f) {
 				logger(LL_ERROR, "%s: fopen failed for %s: %s\n", __func__, filepath, strerror(errno));
 				free(filepath);
-				free(buffer);
 				return -2;
 			}
-			if (fread(buffer, 1, size, f) != size) {
-				fclose(f);
-				logger(LL_ERROR, "%s: fread failed for %s: %s\n", __func__, filepath, strerror(errno));
+			buffer = (unsigned char *)mmap(0, size, PROT_READ, MAP_PRIVATE, fileno(f), 0);
+			fclose(f);
+			if (buffer == MAP_FAILED) {
+				logger(LL_ERROR, "%s: mmap failed for %s: %s\n", __func__, filepath, strerror(errno));
 				free(filepath);
-				free(buffer);
 				return -1;
 			}
-			fclose(f);
 #ifndef WIN32
 		}
 #endif
-		buffer[size] = '\0';
-
 		free(filepath);
 	}
 
@@ -766,8 +767,8 @@ int ipsw_extract_to_memory(ipsw_archive_t ipsw, const char* infile, void** pbuff
 int ipsw_extract_send(ipsw_archive_t ipsw, const char* infile, int blocksize, ipsw_send_cb send_callback, void* ctx)
 {
 	unsigned char* buffer = NULL;
-	size_t done = 0;
-	size_t total_size = 0;
+	uint64_t done = 0;
+	uint64_t total_size = 0;
 
 	if (ipsw == NULL) {
 		logger(LL_ERROR, "Invalid archive\n");
@@ -775,6 +776,9 @@ int ipsw_extract_send(ipsw_archive_t ipsw, const char* infile, int blocksize, ip
 	}
 
 	if (ipsw->zip) {
+		logger(LL_ERROR, "zip mode unsupported with mmap\n");
+		return -1;
+
 		int err = 0;
 		struct zip *zip = zip_open(ipsw->path, 0, &err);
 		if (zip == NULL) {
@@ -851,15 +855,15 @@ int ipsw_extract_send(ipsw_archive_t ipsw, const char* infile, int blocksize, ip
 			return -1;
 		}
 		total_size = fst.st_size;
-		buffer = (unsigned char*)malloc(blocksize);
-		if (buffer == NULL) {
-			logger(LL_ERROR, "Out of memory\n");
-			free(filepath);
-			return -1;
-		}
 
 #ifndef WIN32
 		if (S_ISLNK(fst.st_mode)) {
+			buffer = (unsigned char*)malloc(blocksize);
+			if (buffer == NULL) {
+				logger(LL_ERROR, "Out of memory\n");
+				free(filepath);
+				return -1;
+			}
 			ssize_t rl = readlink(filepath, (char*)buffer, (total_size > blocksize) ? blocksize : total_size);
 			if (rl < 0) {
 				logger(LL_ERROR, "%s: readlink failed for %s: %s\n", __func__, filepath, strerror(errno));
@@ -868,36 +872,37 @@ int ipsw_extract_send(ipsw_archive_t ipsw, const char* infile, int blocksize, ip
 				return -1;
 			}
 			send_callback(ctx, buffer, (size_t)rl, 0, 0);
+			free(buffer);
 		} else {
 #endif
 			FILE *f = fopen(filepath, "rb");
 			if (!f) {
 				logger(LL_ERROR, "%s: fopen failed for %s: %s\n", __func__, filepath, strerror(errno));
 				free(filepath);
-				free(buffer);
 				return -2;
 			}
 
 			while (done < total_size) {
-				size_t size = total_size-done;
+				uint64_t size = total_size - done;
 				if (size > blocksize) size = blocksize;
-				size_t fr = fread(buffer, 1, size, f);
-				if (fr != size) {
-					logger(LL_ERROR, "%s: fread failed for %s: %s\n", __func__, filepath, strerror(errno));
+				buffer = (unsigned char *)mmap(0, size, PROT_READ, MAP_PRIVATE, fileno(f), done);
+				if (buffer == MAP_FAILED) {
+					logger(LL_ERROR, "%s: mmap failed for %s: %s\n", __func__, filepath, strerror(errno));
 					break;
 				}
-				if (send_callback(ctx, buffer, fr, done, total_size) < 0) {
+				if (send_callback(ctx, buffer, size, done, total_size) < 0) {
 					logger(LL_ERROR, "%s: send failed\n", __func__);
+					free_mmap(buffer, size, -1);
 					break;
 				}
-				done += fr;
+				free_mmap(buffer, size, -1);
+				done += size;
 			}
 			fclose(f);
 #ifndef WIN32
 		}
 #endif
 		free(filepath);
-		free(buffer);
 	}
 
 	if (done < total_size) {
@@ -922,7 +927,7 @@ int ipsw_extract_build_manifest(ipsw_archive_t ipsw, plist_t* buildmanifest, int
 	if (ipsw_file_exists(ipsw, "BuildManifesto.plist")) {
 		if (ipsw_extract_to_memory(ipsw, "BuildManifesto.plist", &data, &size) == 0) {
 			plist_from_memory((char*)data, size, buildmanifest, NULL);
-			free(data);
+			free_mmap(data, size, -1);
 			return 0;
 		}
 	}
@@ -934,7 +939,7 @@ int ipsw_extract_build_manifest(ipsw_archive_t ipsw, plist_t* buildmanifest, int
 	if (ipsw_extract_to_memory(ipsw, "BuildManifest.plist", &data, &size) == 0) {
 		*tss_enabled = 1;
 		plist_from_memory((char*)data, size, buildmanifest, NULL);
-		free(data);
+		free_mmap(data, size, -1);
 		return 0;
 	}
 
@@ -948,7 +953,7 @@ int ipsw_extract_restore_plist(ipsw_archive_t ipsw, plist_t* restore_plist)
 
 	if (ipsw_extract_to_memory(ipsw, "Restore.plist", &data, &size) == 0) {
 		plist_from_memory((char*)data, size, restore_plist, NULL);
-		free(data);
+		free_mmap(data, size, -1);
 		return 0;
 	}
 
diff --git a/src/ipsw.h b/src/ipsw.h
index b22e614..65deb38 100644
--- a/src/ipsw.h
+++ b/src/ipsw.h
@@ -44,7 +44,7 @@ void ipsw_close(ipsw_archive_t ipsw);
 int ipsw_print_info(const char* ipsw);
 
 typedef int (*ipsw_list_cb)(void *ctx, ipsw_archive_t ipsw, const char *name, struct stat *stat);
-typedef int (*ipsw_send_cb)(void *ctx, const void *data, size_t size, size_t done, size_t total_size);
+typedef int (*ipsw_send_cb)(void *ctx, const void *data, uint64_t size, uint64_t done, uint64_t total_size);
 
 struct ipsw_file_handle {
 	FILE* file;
diff --git a/src/recovery.c b/src/recovery.c
index b843f1e..f6a8977 100644
--- a/src/recovery.c
+++ b/src/recovery.c
@@ -307,7 +307,7 @@ int recovery_send_component(struct idevicerestore_client_t* client, plist_t buil
 	}
 
 	ret = personalize_component(client, component, component_data, component_size, client->tss, &data, &size);
-	free(component_data);
+	free_mmap(component_data, component_size, -1);
 	if (ret < 0) {
 		logger(LL_ERROR, "Unable to get personalized component: %s\n", component);
 		return -1;
diff --git a/src/restore.c b/src/restore.c
index 04bc7ab..1676f09 100644
--- a/src/restore.c
+++ b/src/restore.c
@@ -1496,7 +1496,7 @@ int restore_send_component(struct idevicerestore_client_t* client, plist_t messa
 	}
 
 	ret = personalize_component(client, component, component_data, component_size, client->tss, &data, &size);
-	free(component_data);
+	free_mmap(component_data, component_size, -1);
 	component_data = NULL;
 	if (ret < 0) {
 		logger(LL_ERROR, "Unable to get personalized component %s\n", component);
@@ -1603,7 +1603,7 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 			snprintf(firmware_filename, sizeof(firmware_filename), "%s/%s", firmware_path, filename);
 			plist_dict_set_item(firmware_files, compname, plist_new_string(firmware_filename));
 		}
-		free(manifest_data);
+		free_mmap(manifest_data, manifest_size, -1);
 	} else {
 		logger(LL_INFO, "Getting firmware manifest from build identity\n");
 		plist_dict_iter iter = NULL;
@@ -1668,7 +1668,7 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 	}
 
 	ret = personalize_component(client, component, component_data, component_size, client->tss, &llb_data, &llb_size);
-	free(component_data);
+	free_mmap(component_data, component_size, -1);
 	component_data = NULL;
 	component_size = 0;
 	if (ret < 0) {
@@ -1727,12 +1727,12 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 			free(iter);
 			free(comp);
 			free(comppath);
-			free(component_data);
+			free_mmap(component_data, component_size, -1);
 			plist_free(firmware_files);
 			logger(LL_ERROR, "Unable to get personalized component: %s\n", component);
 			return -1;
 		}
-		free(component_data);
+		free_mmap(component_data, component_size, -1);
 		component_data = NULL;
 		component_size = 0;
 
@@ -1771,7 +1771,7 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 		}
 
 		ret = personalize_component(client, component, component_data, component_size, client->tss, &personalized_data, &personalized_size);
-		free(component_data);
+		free_mmap(component_data, component_size, -1);
 		component_data = NULL;
 		component_size = 0;
 		if (ret < 0) {
@@ -1796,7 +1796,7 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 		}
 
 		ret = personalize_component(client, component, component_data, component_size, client->tss, &personalized_data, &personalized_size);
-		free(component_data);
+		free_mmap(component_data, component_size, -1);
 		component_data = NULL;
 		component_size = 0;
 		if (ret < 0) {
@@ -1821,7 +1821,7 @@ int restore_send_nor(struct idevicerestore_client_t* client, plist_t message)
 		}
 
 		ret = personalize_component(client, component, component_data, component_size, client->tss, &personalized_data, &personalized_size);
-		free(component_data);
+		free_mmap(component_data, component_size, -1);
 		component_data = NULL;
 		component_size = 0;
 		if (ret < 0) {
@@ -2380,7 +2380,7 @@ static int restore_send_baseband_data(struct idevicerestore_client_t* client, pl
 	// send file
 	dict = plist_new_dict();
 	plist_dict_set_item(dict, "BasebandData", plist_new_data(buffer, sz));
-	free(buffer);
+	free_mmap(buffer, sz, -1);
 	buffer = NULL;
 
 	restore_service_client_t service = _restore_get_service_client_for_data_request(client, message);
@@ -2535,10 +2535,11 @@ static int restore_send_image_data(struct idevicerestore_client_t *client, plist
 						path = NULL;
 						if (ret < 0) {
 							logger(LL_ERROR, "Unable to extract component: %s\n", component);
+							return -1;
 						}
 
 						ret = personalize_component(client, component, component_data, component_size, client->tss, &data, &size);
-						free(component_data);
+						free_mmap(component_data, component_size, -1);
 						component_data = NULL;
 						if (ret < 0) {
 							logger(LL_ERROR, "Unable to get personalized component: %s\n", component);
@@ -2735,7 +2736,7 @@ static plist_t restore_get_se_firmware_data(struct idevicerestore_client_t* clie
 	}
 
 	plist_dict_set_item(response, "FirmwareData", plist_new_data((char*)component_data, component_size));
-	free(component_data);
+	free_mmap(component_data, component_size, NULL);
 	component_data = NULL;
 	component_size = 0;
 
@@ -2821,7 +2822,7 @@ static plist_t restore_get_savage_firmware_data(struct idevicerestore_client_t*
 		return NULL;
 	}
 
-	ret = extract_component(client->ipsw, comp_path, &component_data, &component_size);
+	ret = extract_component(client->ipsw, comp_path, &component_data_tmp, &component_size);
 	free(comp_path);
 	comp_path = NULL;
 	if (ret < 0) {
@@ -2833,14 +2834,14 @@ static plist_t restore_get_savage_firmware_data(struct idevicerestore_client_t*
 	free(comp_name);
 	comp_name = NULL;
 
-	component_data_tmp = realloc(component_data, (size_t)component_size+16);
-	if (!component_data_tmp) {
-		free(component_data);
+	component_data = malloc((size_t)component_size+16);
+	if (!component_data) {
+		free_mmap(component_data_tmp, component_size, NULL);
 		plist_free(response);
 		return NULL;
 	}
-	component_data = component_data_tmp;
-	memmove(component_data + 16, component_data, (size_t)component_size);
+	memcpy(component_data + 16, component_data_tmp, (size_t)component_size);
+	free_mmap(component_data_tmp, component_size, NULL);
 	memset(component_data, '\0', 16);
 	*(uint32_t*)(component_data + 4) = htole32((uint32_t)component_size);
 	component_size += 16;
@@ -2948,7 +2949,7 @@ static plist_t restore_get_yonkers_firmware_data(struct idevicerestore_client_t*
 	plist_dict_set_item(firmware_data, "YonkersFirmware", plist_new_data((char*)component_data, component_size));
 	plist_dict_set_item(response, "FirmwareData", firmware_data);
 
-	free(component_data);
+	free_mmap(component_data, component_size, NULL);
 	component_data = NULL;
 	component_size = 0;
 
@@ -3044,11 +3045,11 @@ static plist_t restore_get_rose_firmware_data(struct idevicerestore_client_t* cl
 	}
 	if (ftab_parse(component_data, component_size, &ftab, &ftag) != 0) {
 		plist_free(response);
-		free(component_data);
+		free_mmap(component_data, component_size, NULL);
 		logger(LL_ERROR, "Failed to parse '%s' component data.\n", comp_name);
 		return NULL;
 	}
-	free(component_data);
+	free_mmap(component_data, component_size, NULL);
 	component_data = NULL;
 	component_size = 0;
 	if (ftag != 'rkos') {
@@ -3075,13 +3076,13 @@ static plist_t restore_get_rose_firmware_data(struct idevicerestore_client_t* cl
 
 		ftag = 0;
 		if (ftab_parse(component_data, component_size, &rftab, &ftag) != 0) {
-			free(component_data);
+			free_mmap(component_data, component_size, NULL);
 			ftab_free(ftab);
 			plist_free(response);
 			logger(LL_ERROR, "Failed to parse '%s' component data.\n", comp_name);
 			return NULL;
 		}
-		free(component_data);
+		free_mmap(component_data, component_size, NULL);
 		component_data = NULL;
 		component_size = 0;
 		if (ftag != 'rkos') {
@@ -3197,7 +3198,7 @@ static plist_t restore_get_veridian_firmware_data(struct idevicerestore_client_t
 	} else {
 		plist_from_xml((const char*)component_data, component_size, &fw_map);
 	}
-	free(component_data);
+	free_mmap(component_data, component_size, NULL);
 	component_data = NULL;
 	component_size = 0;
 
@@ -3366,7 +3367,7 @@ static plist_t restore_get_tcon_firmware_data(struct idevicerestore_client_t* cl
 	}
 
 	plist_dict_set_item(response, "FirmwareData", plist_new_data((char*)component_data, component_size));
-	free(component_data);
+	free_mmap(component_data, component_size, NULL);
 	component_data = NULL;
 	component_size = 0;
 
@@ -3510,12 +3511,12 @@ static plist_t restore_get_timer_firmware_data(struct idevicerestore_client_t* c
 			return NULL;
 		}
 		if (ftab_parse(component_data, component_size, &ftab, &ftag) != 0) {
-			free(component_data);
+			free_mmap(component_data, component_size, NULL);
 			plist_free(response);
 			logger(LL_ERROR, "Failed to parse '%s' component data.\n", comp_name);
 			return NULL;
 		}
-		free(component_data);
+		free_mmap(component_data, component_size, NULL);
 		component_data = NULL;
 		component_size = 0;
 		if (ftag != 'rkos') {
@@ -3545,13 +3546,13 @@ static plist_t restore_get_timer_firmware_data(struct idevicerestore_client_t* c
 
 		ftag = 0;
 		if (ftab_parse(component_data, component_size, &rftab, &ftag) != 0) {
-			free(component_data);
+			free_mmap(component_data, component_size, NULL);
 			ftab_free(ftab);
 			plist_free(response);
 			logger(LL_ERROR, "Failed to parse '%s' component data.\n", comp_name);
 			return NULL;
 		}
-		free(component_data);
+		free_mmap(component_data, component_size, NULL);
 		component_data = NULL;
 		component_size = 0;
 		if (ftag != 'rkos') {
@@ -4024,9 +4025,9 @@ static int restore_bootability_send_one(void *ctx, ipsw_archive_t ipsw, const ch
 
 	if ((S_ISLNK(stat->st_mode) || S_ISREG(stat->st_mode)) && stat->st_size != 0) {
 		ipsw_extract_to_memory(ipsw, name, &buf, &size);
-		if (size != stat->st_size) {
+		if (!S_ISLNK(stat->st_mode) && size != stat->st_size) {
 			logger(LL_ERROR, "expected %zu bytes but got %zu for file %s\n", (size_t)stat->st_size, size, name);
-			free(buf);
+			free_mmap(buf, size, -1);
 			return -1;
 		}
 	}
@@ -4035,7 +4036,7 @@ static int restore_bootability_send_one(void *ctx, ipsw_archive_t ipsw, const ch
 
 	int ret = cpio_send_file(connection, subpath, stat, buf);
 
-	free(buf);
+	free_mmap(buf, size, -1);
 	return ret;
 }
 
@@ -4220,7 +4221,7 @@ struct _restore_send_file_data_ctx {
 	uint32_t tag;
 };
 
-static int _restore_send_file_data(struct _restore_send_file_data_ctx* rctx, const void* data, size_t size, size_t done, size_t total_size)
+static int _restore_send_file_data(struct _restore_send_file_data_ctx* rctx, const void* data, uint64_t size, uint64_t done, uint64_t total_size)
 {
 	plist_t dict = plist_new_dict();
 	if (data != NULL) {
@@ -4290,21 +4291,25 @@ int restore_send_personalized_boot_object_v3(struct idevicerestore_client_t* cli
 	size_t size = 0;
 	void *data = NULL;
 	char *path = NULL;
+	bool is_mmap = false;
 
 	logger(LL_INFO, "About to send %s...\n", component);
 
 	if (strcmp(image_name, "__GlobalManifest__") == 0) {
+		is_mmap = true;
 		int ret = extract_global_manifest(client, client->restore->build_identity, NULL, &data, &size);
 		if (ret != 0) {
 			return -1;
 		}
 	} else if (strcmp(image_name, "__RestoreVersion__") == 0) {
+		is_mmap = true;
 		int ret = ipsw_extract_to_memory(client->ipsw, "RestoreVersion.plist", &data, &size);
 		if (ret != 0) {
 			logger(LL_ERROR, "failed to read global manifest\n");
 			return -1;
 		}
 	} else if (strcmp(image_name, "__SystemVersion__") == 0) {
+		is_mmap = true;
 		int ret = ipsw_extract_to_memory(client->ipsw, "SystemVersion.plist", &data, &size);
 		if (ret != 0) {
 			logger(LL_ERROR, "failed to read global manifest\n");
@@ -4342,7 +4347,7 @@ int restore_send_personalized_boot_object_v3(struct idevicerestore_client_t* cli
 
 		// Personalize IMG4
 		ret = personalize_component(client, component, component_data, component_size, client->tss, &data, &size);
-		free(component_data);
+		free_mmap(component_data, component_size, -1);
 		component_data = NULL;
 		if (ret < 0) {
 			logger(LL_ERROR, "Unable to get personalized component %s\n", component);
@@ -4353,6 +4358,11 @@ int restore_send_personalized_boot_object_v3(struct idevicerestore_client_t* cli
 	restore_service_client_t service = _restore_get_service_client_for_data_request(client, message);
 	if (!service) {
 		logger(LL_ERROR, "%s: Unable to connect to service client\n", __func__);
+		if (is_mmap) {
+			free_mmap(data, size, -1);
+		} else {
+			free(data);
+		}
 		return -1;
 	}
 
@@ -4372,7 +4382,11 @@ int restore_send_personalized_boot_object_v3(struct idevicerestore_client_t* cli
 	while (i > 0) {
 		int blob_size = i > 8192 ? 8192 : i;
 		if (_restore_send_file_data(&rctx, ((char*)data + size - i), blob_size, size-i, size) < 0) {
-			free(data);
+			if (is_mmap) {
+				free_mmap(data, size, -1);
+			} else {
+				free(data);
+			}
 			_restore_service_free(service);
 			finalize_progress(rctx.tag);
 			logger(LL_ERROR, "Unable to send component %s data\n", component);
@@ -4380,7 +4394,11 @@ int restore_send_personalized_boot_object_v3(struct idevicerestore_client_t* cli
 		}
 		i -= blob_size;
 	}
-	free(data);
+	if (is_mmap) {
+		free_mmap(data, size, -1);
+	} else {
+		free(data);
+	}
 
 	_restore_send_file_data(&rctx, NULL, 0, size-i, size);
 	finalize_progress(rctx.tag);
@@ -4689,7 +4707,7 @@ int restore_send_recovery_os_file_asset_image(struct idevicerestore_client_t* cl
 	void* data = NULL;
 	size_t size = 0;
 	ret = personalize_component(client, component, component_data, component_size, client->tss_recoveryos_root_ticket, &data, &size);
-	free(component_data);
+	free_mmap(component_data, component_size, -1);
 	component_data = NULL;
 	if (ret < 0) {
 		logger(LL_ERROR, "Unable to get personalized component %s. Trying to continue anyway.\n", component);
@@ -4752,7 +4770,7 @@ int restore_send_recovery_os_iboot_fw_files_images(struct idevicerestore_client_
 							void* data = NULL;
 							size_t size = 0;
 							ret = personalize_component(client, component, component_data, component_size, client->tss_recoveryos_root_ticket, &data, &size);
-							free(component_data);
+							free_mmap(component_data, component_size, -1);
 							component_data = NULL;
 							if (ret == 0) {
 								plist_dict_set_item(firmware_files, component, plist_new_data((char*)data, size));
